light-client/src/builder/light_client.rs

```diff
@@ -0,0 +1,213 @@
+    crate::components::scheduler, crate::components::verifier::ProdVerifier,
+    crate::operations::ProdHasher, crate::predicates::ProdPredicates, std::time::Duration,
+    tendermint_rpc as rpc,
+};
+
+/// No trusted state has been set yet
+pub struct NoTrustedState;
+
+/// A trusted state has been set and validated
+pub struct HasTrustedState;
+
+/// Builder for a light client [`Instance`]
+#[must_use]
+pub struct LightClientBuilder<State> {
+    peer_id: PeerId,
+    options: Options,
+    io: Box<dyn Io>,
+    clock: Box<dyn Clock>,
+    hasher: Box<dyn Hasher>,
+    verifier: Box<dyn Verifier>,
+    scheduler: Box<dyn Scheduler>,
+    predicates: Box<dyn VerificationPredicates>,
+    light_store: Box<dyn LightStore>,
+
+    #[allow(dead_code)]
+    state: State,
+}
+
+impl<Current> LightClientBuilder<Current> {
+    /// Private method to move from one state to another
+    fn with_state<Next>(self, state: Next) -> LightClientBuilder<Next> {
+        LightClientBuilder {
+            peer_id: self.peer_id,
+            options: self.options,
+            io: self.io,
+            clock: self.clock,
+            hasher: self.hasher,
+            verifier: self.verifier,
+            scheduler: self.scheduler,
+            predicates: self.predicates,
+            light_store: self.light_store,
+            state,
+        }
+    }
+}
+
+impl LightClientBuilder<NoTrustedState> {
+    /// Initialize a builder for a production (non-mock) light client.
+    #[cfg(feature = "rpc-client")]
+    pub fn prod(
+        peer_id: PeerId,
+        rpc_client: rpc::HttpClient,
+        light_store: Box<dyn LightStore>,
+        options: Options,
+        timeout: Option<Duration>,
+    ) -> Self {
+        Self::custom(
+            peer_id,
+            options,
+            light_store,
+            Box::new(ProdIo::new(peer_id, rpc_client, timeout)),
+            Box::new(ProdHasher),
+            Box::new(SystemClock),
+            Box::new(ProdVerifier::default()),
+            Box::new(scheduler::basic_bisecting_schedule),
+            Box::new(ProdPredicates),
+        )
+    }
+
+    /// Initialize a builder for a custom light client, by providing all dependencies upfront.
+    #[allow(clippy::too_many_arguments)]
+    pub fn custom(
+        peer_id: PeerId,
+        options: Options,
+        light_store: Box<dyn LightStore>,
+        io: Box<dyn Io>,
+        hasher: Box<dyn Hasher>,
+        clock: Box<dyn Clock>,
+        verifier: Box<dyn Verifier>,
+        scheduler: Box<dyn Scheduler>,
+        predicates: Box<dyn VerificationPredicates>,
+    ) -> Self {
+        Self {
+            peer_id,
+            hasher,
+            io,
+            verifier,
+            light_store,
+            clock,
+            scheduler,
+            options,
+            predicates,
+            state: NoTrustedState,
+        }
+    }
+
+    /// Set the given light block as the initial trusted state.
+    pub fn trust_light_block(
```

Comment:
Should this be public?

Reply:
Good catch. It should not, as it is quite dangerous and is only currently used in the light client example. The latter could be rewritten to use `trust_from_store` instead so it's not needed at all. Will remove it.

Reply:
2705247